transcom/mymove
# [MB-13333] Add backend file upload support for PPM document spreadsheets

## [Jira ticket](https://dp3.atlassian.net/browse/MB-13333) for this change

## Summary

PPMs have added a new requirement that we allow spreadsheet files to be uploaded when a weight ticket or pro-gear closeout document is used instead of an image or pdf of the supporting proof.

We made the frontend client changes to our upload component previously but our backend code wasn't yet modified to allow and detect spreadsheet file types.

I added content types for `.xls` and `.xlsx` on the backend to match what we have on the client side but our use of the standard library [DetectContentType](https://pkg.go.dev/net/http#DetectContentType) would always return `application/octet-stream` for excel uploads (or other non-supported file types).  This is because the standard isn't supposed to be a robust MIME type check but is more used for limited API request purposes.  I went looking and found a fuller featured library and found a good option https://github.com/gabriel-vasile/mimetype.

The other problem I was trying to address once files were being accurately identified, was to have our PPM specific pages use a new endpoint separate from the one being shared with uploading orders for the customer (that should only allow images and PDFs).  While it was nice that we could reuse this endpoint to this point, we would have made it more complex by trying to verify which page was making the request and manage the different business rules.

[Transcom Weight Estimator download](https://www.ustranscom.mil/dp3/weightestimator.cfm)

https://dp3.atlassian.net/wiki/spaces/MT/pages/1470955567/How+to+test+storing+data+in+S3+locally

## Setup to Run Your Code

To test using local storage:

`make server_generate && make server_run`

`make client_run`

To test using AWS S3 storage:

Uncomment the line in your .envrc file:
`export STORAGE_BACKEND=s3`

`direnv reload`
`aws-vault exec transcom-gov-dev -- make server_run`

To login to AWS gov-dev and see the S3 uploads you can run `aws-vault login transcom-gov-dev` and select the S3 service.  You'll find the uploads under the `user/` prefix for the user you logged into the customer app as.

`make client_run`

### Additional steps



1. Login to the customer app with a PPM with an existing weight ticket or at least a PPM ready for closeout like `weightTicketPPM@ppm.approved`
2. Proceed to choose Upload PPM Documents from the home screen
3. Add a new weight ticket or edit an existing weight ticket of the PPM
4. In either of the empty weight or full weight upload fields you should be able to select one of the test files [Weight Estimator.xls](https://github.com/transcom/mymove/blob/master/pkg/testdatagen/testdata/Weight%20Estimator.xls) or [Weight Estimator.xlsx](https://github.com/transcom/mymove/blob/master/pkg/testdatagen/testdata/Weight%20Estimator.xlsx)
5. It's also good to verify that behavior hasn't changed for uploading orders docs or image and pdf documents since they are using the new content detection library.

## Verification Steps for Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.



### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots

**Tested on Experimental**

![Screen Shot 2022-10-17 at 1 28 12 PM](https://user-images.githubusercontent.com/52669884/196280721-0ff20ea4-c9c8-44c4-8e79-ec6a275dbcd8.png)

**Inspect the content type saved to the uploads database table**
<img width="927" alt="Screen Shot 2022-10-17 at 5 10 22 PM" src="https://user-images.githubusercontent.com/52669884/196284158-d686a226-2c61-4fc6-8ab0-875bb6ecc046.png">
